src/core/observables/RDF.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2010-2020 The ESPResSo project
+ *
+ * This file is part of ESPResSo.
+ *
+ * ESPResSo is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * ESPResSo is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "RDF.hpp"
+
+#include "fetch_particles.hpp"
+#include "particle_data.hpp"
+
+#include <utils/math/int_pow.hpp>
+
+#include <boost/range/algorithm/transform.hpp>
+#include <cmath>
+
+namespace Observables {
+std::vector<double> RDF::operator()() const {
+  std::vector<Particle> particles1 = fetch_particles(ids1());
+  std::vector<const Particle *> particles_ptrs1(particles1.size());
+  boost::transform(particles1, particles_ptrs1.begin(),
+                   [](auto const &p) { return std::addressof(p); });
+
+  if (!ids2().empty()) {
+    std::vector<Particle> particles2 = fetch_particles(ids2());
+    std::vector<const Particle *> particles_ptrs2(particles2.size());
+    boost::transform(particles2, particles_ptrs2.begin(),
+                     [](auto const &p) { return std::addressof(p); });
+    return this->evaluate(particles_ptrs1, particles_ptrs2);
+  }
+
+  return this->evaluate(particles_ptrs1, {});
+}
+
+std::vector<double>
+RDF::evaluate(Utils::Span<const Particle *const> particles1,
+              Utils::Span<const Particle *const> particles2) const {
+  auto const bin_width = (max_r - min_r) / static_cast<double>(n_r_bins);
+  auto const inv_bin_width = 1.0 / bin_width;
+  std::vector<double> res(n_values(), 0.0);
+  long int cnt = 0;
+  bool const mixed_flag = !ids2().empty();
+  for (auto it = particles1.begin(); it != particles1.end(); ++it) {
+    for (auto jt = mixed_flag ? particles2.begin() : std::next(it),
+              jend = mixed_flag ? particles2.end() : particles1.end();
+         jt != jend; ++jt) {
+      auto const p1 = *it, p2 = *jt;
+      auto const dist = get_mi_vector(p1->r.p, p2->r.p, box_geo).norm();
+      if (dist > min_r && dist < max_r) {
+        auto const ind =
+            static_cast<int>(std::floor((dist - min_r) * inv_bin_width));
+        res[ind]++;
+      }
```

Comment:
I think we once said that we want to store particle pairs that should be considered. I think the creation of the pairs should not be part of the observable here. What do you think?

Reply:
It didn't cross my mind. I've tinkered with the RDF code, but it's not immediately obvious to me how to delegate pair creation to another function. If we store a `std::vector<std::pair<int,int>> particle_pairs` in the RDF class, we'll have to copy either `N*M` or `N*N/2` particle objects instead of `N+M` resp. `N` particle objects. If we delegate particle pair creation to the `operator()()` method, then we only move the logic to another member and add an extra overhead (3x slowdown in f71695c0f3c). Do you have a better strategy in mind?

Reply:
I don't think explicitly passing the pairs is feasible (as you already found out). More abstractly speaking, what the RDF is calculating is, given two ranges, an folding (or reduction) over the cartesian product of two ranges plus an reduction kernel. The first one is a general thing that can be reused (we already have a special case of this, `Utils::for_each_pair`, which is the special case where both ranges are the same), and the kernel is calculating the distance, and updating a histogram. So in the end I think for a proper design we have to do the same separation between kernels and accumulation for pairs as we did for the single particle observables, but perhaps with slightly different algorithms (?).

Reply:
I've just created an overload for `Utils::for_each_pair()` that takes two ranges in a052ee8, and `RDF::evaluate()` now runs 5% faster in the mixed case, 1% slower in the unmixed case. Could still be improved, e.g. the overload could decay to the original if both range iterators are the same.

Reply:
I think we should go forward with this, this is a good starting point for latter refactorings. I have some comments on your implementation of `for_each_pair`, but we can discuss that in the actual review. (There would be different possibilities for the design, e.g. you could implement a meta-range that runs over the cartesian products, but I think this is clearer and more "STL-like" so potentially less surprising). I think if we properly dispatch to the single rang e case in the implementation of `for_each_pair` we should be able to obtain at least the same performance as the old code...

Reply:
Alright, the overload was a bad idea due to template collision, had to find a different name. Concerning performance, the timing actually changes whenever the project is recompiled, so I can't really get something consistent, it's more or less within 1% compared to the hardcoded pair loop. I don't think it matters that much, the hardcoded pair loop implementation was already 10% and 25% faster in the mixed and unmixed case compared to the original `statistics.cpp` implementation.

Reply:
I agree, let's not worry too much about the performance. The algorithm can later be optimized in isolation...